File: api/v1/endpoints/hr_agricole.py
"""
Endpoints API pour la gestion des fonctionnalités RH agricoles
"""

from datetime import date
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from db.database import get_db
from services.hr_agricole_service import HRAgricoleService
from schemas.hr_agricole import (
    CompetenceAgricoleCreate, CompetenceAgricoleUpdate, CompetenceAgricoleInDB,
    CertificationAgricoleCreate, CertificationAgricoleUpdate, CertificationAgricoleInDB,
    AffectationParcelleCreate, AffectationParcelleUpdate, AffectationParcelleInDB,
    ConditionTravailAgricoleCreate, ConditionTravailAgricoleUpdate, ConditionTravailAgricoleInDB,
    FormationAgricoleCreate, FormationAgricoleUpdate, FormationAgricoleInDB,
    EvaluationAgricoleCreate, EvaluationAgricoleUpdate, EvaluationAgricoleInDB
)

router = APIRouter(prefix="/hr-agricole", tags=["RH Agricole"])

# Dépendances
def get_hr_agricole_service(db: Session = Depends(get_db)) -> HRAgricoleService:
    return HRAgricoleService(db)

# Endpoints Compétences
@router.post("/competences/", response_model=CompetenceAgricoleInDB, status_code=status.HTTP_201_CREATED)
def create_competence(
    competence: CompetenceAgricoleCreate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Crée une nouvelle compétence agricole"""
    return service.create_competence(competence)

@router.get("/competences/{competence_id}", response_model=CompetenceAgricoleInDB)
def get_competence(
    competence_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère une compétence par son ID"""
    competence = service.get_competence(competence_id)
    if not competence:
        raise HTTPException(status_code=404, detail="Compétence non trouvée")
    return competence

@router.get("/employes/{employe_id}/competences/", response_model=List[CompetenceAgricoleInDB])
def get_competences_employe(
    employe_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère toutes les compétences d'un employé"""
    return service.get_competences_employe(employe_id)

@router.put("/competences/{competence_id}", response_model=CompetenceAgricoleInDB)
def update_competence(
    competence_id: str,
    competence: CompetenceAgricoleUpdate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Met à jour une compétence"""
    updated = service.update_competence(competence_id, competence)
    if not updated:
        raise HTTPException(status_code=404, detail="Compétence non trouvée")
    return updated

# Endpoints Certifications
@router.post("/certifications/", response_model=CertificationAgricoleInDB, status_code=status.HTTP_201_CREATED)
def create_certification(
    certification: CertificationAgricoleCreate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Crée une nouvelle certification agricole"""
    return service.create_certification(certification)

@router.get("/certifications/{certification_id}", response_model=CertificationAgricoleInDB)
def get_certification(
    certification_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère une certification par son ID"""
    certification = service.get_certification(certification_id)
    if not certification:
        raise HTTPException(status_code=404, detail="Certification non trouvée")
    return certification

@router.get("/competences/{competence_id}/certifications/", response_model=List[CertificationAgricoleInDB])
def get_certifications_competence(
    competence_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère toutes les certifications d'une compétence"""
    return service.get_certifications_competence(competence_id)

@router.put("/certifications/{certification_id}", response_model=CertificationAgricoleInDB)
def update_certification(
    certification_id: str,
    certification: CertificationAgricoleUpdate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Met à jour une certification"""
    updated = service.update_certification(certification_id, certification)
    if not updated:
        raise HTTPException(status_code=404, detail="Certification non trouvée")
    return updated

# Endpoints Affectations
@router.post("/affectations/", response_model=AffectationParcelleInDB, status_code=status.HTTP_201_CREATED)
def create_affectation(
    affectation: AffectationParcelleCreate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Crée une nouvelle affectation"""
    return service.create_affectation(affectation)

@router.get("/affectations/{affectation_id}", response_model=AffectationParcelleInDB)
def get_affectation(
    affectation_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère une affectation par son ID"""
    affectation = service.get_affectation(affectation_id)
    if not affectation:
        raise HTTPException(status_code=404, detail="Affectation non trouvée")
    return affectation

@router.get("/employes/{employe_id}/affectations/", response_model=List[AffectationParcelleInDB])
def get_affectations_employe(
    employe_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère toutes les affectations d'un employé"""
    return service.get_affectations_employe(employe_id)

@router.get("/parcelles/{parcelle_id}/affectations/", response_model=List[AffectationParcelleInDB])
def get_affectations_parcelle(
    parcelle_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère toutes les affectations d'une parcelle"""
    return service.get_affectations_parcelle(parcelle_id)

@router.put("/affectations/{affectation_id}", response_model=AffectationParcelleInDB)
def update_affectation(
    affectation_id: str,
    affectation: AffectationParcelleUpdate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Met à jour une affectation"""
    updated = service.update_affectation(affectation_id, affectation)
    if not updated:
        raise HTTPException(status_code=404, detail="Affectation non trouvée")
    return updated

# Endpoints Conditions de travail
@router.post("/conditions-travail/", response_model=ConditionTravailAgricoleInDB, status_code=status.HTTP_201_CREATED)
def create_condition_travail(
    condition: ConditionTravailAgricoleCreate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Crée une nouvelle condition de travail"""
    return service.create_condition_travail(condition)

@router.get("/conditions-travail/{condition_id}", response_model=ConditionTravailAgricoleInDB)
def get_condition_travail(
    condition_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère une condition de travail par son ID"""
    condition = service.get_condition_travail(condition_id)
    if not condition:
        raise HTTPException(status_code=404, detail="Condition de travail non trouvée")
    return condition

@router.get("/employes/{employe_id}/conditions-travail/", response_model=List[ConditionTravailAgricoleInDB])
def get_conditions_travail_employe(
    employe_id: str,
    date_debut: date,
    date_fin: date,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère les conditions de travail d'un employé sur une période"""
    return service.get_conditions_travail_employe(employe_id, date_debut, date_fin)

@router.put("/conditions-travail/{condition_id}", response_model=ConditionTravailAgricoleInDB)
def update_condition_travail(
    condition_id: str,
    condition: ConditionTravailAgricoleUpdate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Met à jour une condition de travail"""
    updated = service.update_condition_travail(condition_id, condition)
    if not updated:
        raise HTTPException(status_code=404, detail="Condition de travail non trouvée")
    return updated

# Endpoints Formations agricoles
@router.post("/formations-agricoles/", response_model=FormationAgricoleInDB, status_code=status.HTTP_201_CREATED)
def create_formation_agricole(
    formation: FormationAgricoleCreate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Crée une nouvelle formation agricole"""
    return service.create_formation_agricole(formation)

@router.get("/formations-agricoles/{formation_id}", response_model=FormationAgricoleInDB)
def get_formation_agricole(
    formation_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère une formation agricole par son ID"""
    formation = service.get_formation_agricole(formation_id)
    if not formation:
        raise HTTPException(status_code=404, detail="Formation agricole non trouvée")
    return formation

@router.get("/formations/{formation_id}/details-agricoles/", response_model=List[FormationAgricoleInDB])
def get_formations_agricoles_formation(
    formation_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère les détails agricoles d'une formation"""
    return service.get_formations_agricoles_formation(formation_id)

@router.put("/formations-agricoles/{formation_id}", response_model=FormationAgricoleInDB)
def update_formation_agricole(
    formation_id: str,
    formation: FormationAgricoleUpdate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Met à jour une formation agricole"""
    updated = service.update_formation_agricole(formation_id, formation)
    if not updated:
        raise HTTPException(status_code=404, detail="Formation agricole non trouvée")
    return updated

# Endpoints Évaluations agricoles
@router.post("/evaluations-agricoles/", response_model=EvaluationAgricoleInDB, status_code=status.HTTP_201_CREATED)
def create_evaluation_agricole(
    evaluation: EvaluationAgricoleCreate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Crée une nouvelle évaluation agricole"""
    return service.create_evaluation_agricole(evaluation)

@router.get("/evaluations-agricoles/{evaluation_id}", response_model=EvaluationAgricoleInDB)
def get_evaluation_agricole(
    evaluation_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère une évaluation agricole par son ID"""
    evaluation = service.get_evaluation_agricole(evaluation_id)
    if not evaluation:
        raise HTTPException(status_code=404, detail="Évaluation agricole non trouvée")
    return evaluation

@router.get("/evaluations/{evaluation_id}/details-agricoles/", response_model=List[EvaluationAgricoleInDB])
def get_evaluations_agricoles_evaluation(
    evaluation_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère les détails agricoles d'une évaluation"""
    return service.get_evaluations_agricoles_evaluation(evaluation_id)

@router.put("/evaluations-agricoles/{evaluation_id}", response_model=EvaluationAgricoleInDB)
def update_evaluation_agricole(
    evaluation_id: str,
    evaluation: EvaluationAgricoleUpdate,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Met à jour une évaluation agricole"""
    updated = service.update_evaluation_agricole(evaluation_id, evaluation)
    if not updated:
        raise HTTPException(status_code=404, detail="Évaluation agricole non trouvée")
    return updated

# Endpoints utilitaires
@router.get("/competences/{competence_id}/validite/")
def check_competence_validite(
    competence_id: str,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Vérifie si une compétence est toujours valide"""
    return {"valide": service.check_competence_validite(competence_id)}

@router.get("/competences/a-renouveler/", response_model=List[CompetenceAgricoleInDB])
def get_competences_a_renouveler(
    delai_jours: int = 30,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère les compétences à renouveler dans un délai donné"""
    return service.get_competences_a_renouveler(delai_jours)

@router.get("/certifications/a-renouveler/", response_model=List[CertificationAgricoleInDB])
def get_certifications_a_renouveler(
    delai_jours: int = 30,
    service: HRAgricoleService = Depends(get_hr_agricole_service)
):
    """Récupère les certifications à renouveler dans un délai donné"""
    return service.get_certifications_a_renouveler(delai_jours)

File: frontend/src/services/hr_agricole.ts
import { api } from '@config/axios';
import {
  CompetenceAgricole,
  CompetenceAgricoleCreate,
  CompetenceAgricoleUpdate,
  CertificationAgricole,
  CertificationAgricoleCreate,
  CertificationAgricoleUpdate,
  AffectationParcelle,
  AffectationParcelleCreate,
  AffectationParcelleUpdate,
  ConditionTravailAgricole,
  ConditionTravailAgricoleCreate,
  ConditionTravailAgricoleUpdate,
  FormationAgricole,
  FormationAgricoleCreate,
  FormationAgricoleUpdate,
  EvaluationAgricole,
  EvaluationAgricoleCreate,
  EvaluationAgricoleUpdate,
  ValidationCompetence,
  CompetenceARenouveler,
  CertificationARenouveler,
  FiltresRHAgricole,
  OptionsTriRHAgricole,
  PaginationRHAgricole,
  StatistiquesCompetences,
  StatistiquesCertifications,
  StatistiquesFormations,
  StatistiquesConditionsTravail
} from '../types/hr_agricole';
import { UUID } from '../types/common';

const BASE_URL = '/hr-agricole';

// Compétences
export const createCompetence = async (competence: CompetenceAgricoleCreate): Promise<CompetenceAgricole> => {
  const response = await api.post<CompetenceAgricole>(`${BASE_URL}/competences/`, competence);
  return response.data;
};

export const getCompetence = async (id: UUID): Promise<CompetenceAgricole> => {
  const response = await api.get<CompetenceAgricole>(`${BASE_URL}/competences/${id}`);
  return response.data;
};

export const getCompetencesEmploye = async (employeId: UUID): Promise<CompetenceAgricole[]> => {
  const response = await api.get<CompetenceAgricole[]>(`${BASE_URL}/employes/${employeId}/competences/`);
  return response.data;
};

export const updateCompetence = async (id: UUID, competence: CompetenceAgricoleUpdate): Promise<CompetenceAgricole> => {
  const response = await api.put<CompetenceAgricole>(`${BASE_URL}/competences/${id}`, competence);
  return response.data;
};

// Certifications
export const createCertification = async (certification: CertificationAgricoleCreate): Promise<CertificationAgricole> => {
  const response = await api.post<CertificationAgricole>(`${BASE_URL}/certifications/`, certification);
  return response.data;
};

export const getCertification = async (id: UUID): Promise<CertificationAgricole> => {
  const response = await api.get<CertificationAgricole>(`${BASE_URL}/certifications/${id}`);
  return response.data;
};

export const getCertificationsCompetence = async (competenceId: UUID): Promise<CertificationAgricole[]> => {
  const response = await api.get<CertificationAgricole[]>(`${BASE_URL}/competences/${competenceId}/certifications/`);
  return response.data;
};

export const updateCertification = async (id: UUID, certification: CertificationAgricoleUpdate): Promise<CertificationAgricole> => {
  const response = await api.put<CertificationAgricole>(`${BASE_URL}/certifications/${id}`, certification);
  return response.data;
};

// Affectations
export const createAffectation = async (affectation: AffectationParcelleCreate): Promise<AffectationParcelle> => {
  const response = await api.post<AffectationParcelle>(`${BASE_URL}/affectations/`, affectation);
  return response.data;
};

export const getAffectation = async (id: UUID): Promise<AffectationParcelle> => {
  const response = await api.get<AffectationParcelle>(`${BASE_URL}/affectations/${id}`);
  return response.data;
};

export const getAffectationsEmploye = async (employeId: UUID): Promise<AffectationParcelle[]> => {
  const response = await api.get<AffectationParcelle[]>(`${BASE_URL}/employes/${employeId}/affectations/`);
  return response.data;
};

export const getAffectationsParcelle = async (parcelleId: UUID): Promise<AffectationParcelle[]> => {
  const response = await api.get<AffectationParcelle[]>(`${BASE_URL}/parcelles/${parcelleId}/affectations/`);
  return response.data;
};

export const updateAffectation = async (id: UUID, affectation: AffectationParcelleUpdate): Promise<AffectationParcelle> => {
  const response = await api.put<AffectationParcelle>(`${BASE_URL}/affectations/${id}`, affectation);
  return response.data;
};

// Conditions de travail
export const createConditionTravail = async (condition: ConditionTravailAgricoleCreate): Promise<ConditionTravailAgricole> => {
  const response = await api.post<ConditionTravailAgricole>(`${BASE_URL}/conditions-travail/`, condition);
  return response.data;
};

export const getConditionTravail = async (id: UUID): Promise<ConditionTravailAgricole> => {
  const response = await api.get<ConditionTravailAgricole>(`${BASE_URL}/conditions-travail/${id}`);
  return response.data;
};

export const getConditionsTravailEmploye = async (
  employeId: UUID,
  dateDebut: string,
  dateFin: string
): Promise<ConditionTravailAgricole[]> => {
  const response = await api.get<ConditionTravailAgricole[]>(
    `${BASE_URL}/employes/${employeId}/conditions-travail/`,
    { params: { date_debut: dateDebut, date_fin: dateFin } }
  );
  return response.data;
};

export const updateConditionTravail = async (
  id: UUID,
  condition: ConditionTravailAgricoleUpdate
): Promise<ConditionTravailAgricole> => {
  const response = await api.put<ConditionTravailAgricole>(`${BASE_URL}/conditions-travail/${id}`, condition);
  return response.data;
};

// Formations agricoles
export const createFormationAgricole = async (formation: FormationAgricoleCreate): Promise<FormationAgricole> => {
  const response = await api.post<FormationAgricole>(`${BASE_URL}/formations-agricoles/`, formation);
  return response.data;
};

export const getFormationAgricole = async (id: UUID): Promise<FormationAgricole> => {
  const response = await api.get<FormationAgricole>(`${BASE_URL}/formations-agricoles/${id}`);
  return response.data;
};

export const getFormationsAgricolesFormation = async (formationId: UUID): Promise<FormationAgricole[]> => {
  const response = await api.get<FormationAgricole[]>(`${BASE_URL}/formations/${formationId}/details-agricoles/`);
  return response.data;
};

export const updateFormationAgricole = async (
  id: UUID,
  formation: FormationAgricoleUpdate
): Promise<FormationAgricole> => {
  const response = await api.put<FormationAgricole>(`${BASE_URL}/formations-agricoles/${id}`, formation);
  return response.data;
};

// Évaluations agricoles
export const createEvaluationAgricole = async (evaluation: EvaluationAgricoleCreate): Promise<EvaluationAgricole> => {
  const response = await api.post<EvaluationAgricole>(`${BASE_URL}/evaluations-agricoles/`, evaluation);
  return response.data;
};

export const getEvaluationAgricole = async (id: UUID): Promise<EvaluationAgricole> => {
  const response = await api.get<EvaluationAgricole>(`${BASE_URL}/evaluations-agricoles/${id}`);
  return response.data;
};

export const getEvaluationsAgricolesEvaluation = async (evaluationId: UUID): Promise<EvaluationAgricole[]> => {
  const response = await api.get<EvaluationAgricole[]>(`${BASE_URL}/evaluations/${evaluationId}/details-agricoles/`);
  return response.data;
};

export const updateEvaluationAgricole = async (
  id: UUID,
  evaluation: EvaluationAgricoleUpdate
): Promise<EvaluationAgricole> => {
  const response = await api.put<EvaluationAgricole>(`${BASE_URL}/evaluations-agricoles/${id}`, evaluation);
  return response.data;
};

// Utilitaires
export const checkCompetenceValidite = async (competenceId: UUID): Promise<ValidationCompetence> => {
  const response = await api.get<ValidationCompetence>(`${BASE_URL}/competences/${competenceId}/validite/`);
  return response.data;
};

export const getCompetencesARenouveler = async (delaiJours: number = 30): Promise<CompetenceARenouveler[]> => {
  const response = await api.get<CompetenceARenouveler[]>(`${BASE_URL}/competences/a-renouveler/`, {
    params: { delai_jours: delaiJours }
  });
  return response.data;
};

export const getCertificationsARenouveler = async (delaiJours: number = 30): Promise<CertificationARenouveler[]> => {
  const response = await api.get<CertificationARenouveler[]>(`${BASE_URL}/certifications/a-renouveler/`, {
    params: { delai_jours: delaiJours }
  });
  return response.data;
};

// Statistiques
export const getStatistiquesCompetences = async (): Promise<StatistiquesCompetences> => {
  const response = await api.get<StatistiquesCompetences>(`${BASE_URL}/statistiques/competences/`);
  return response.data;
};

export const getStatistiquesCertifications = async (): Promise<StatistiquesCertifications> => {
  const response = await api.get<StatistiquesCertifications>(`${BASE_URL}/statistiques/certifications/`);
  return response.data;
};

export const getStatistiquesFormations = async (): Promise<StatistiquesFormations> => {
  const response = await api.get<StatistiquesFormations>(`${BASE_URL}/statistiques/formations/`);
  return response.data;
};

export const getStatistiquesConditionsTravail = async (
  dateDebut: string,
  dateFin: string
): Promise<StatistiquesConditionsTravail> => {
  const response = await api.get<StatistiquesConditionsTravail>(`${BASE_URL}/statistiques/conditions-travail/`, {
    params: { date_debut: dateDebut, date_fin: dateFin }
  });
  return response.data;
};

// Recherche et filtrage
export const rechercherCompetences = async (
  filtres: FiltresRHAgricole,
  tri: OptionsTriRHAgricole,
  pagination: PaginationRHAgricole
): Promise<{ data: CompetenceAgricole[]; pagination: PaginationRHAgricole }> => {
  const response = await api.get<{ data: CompetenceAgricole[]; pagination: PaginationRHAgricole }>(
    `${BASE_URL}/competences/recherche/`,
    { params: { ...filtres, ...tri, ...pagination } }
  );
  return response.data;
};

export const rechercherCertifications = async (
  filtres: FiltresRHAgricole,
  tri: OptionsTriRHAgricole,
  pagination: PaginationRHAgricole
): Promise<{ data: CertificationAgricole[]; pagination: PaginationRHAgricole }> => {
  const response = await api.get<{ data: CertificationAgricole[]; pagination: PaginationRHAgricole }>(
    `${BASE_URL}/certifications/recherche/`,
    { params: { ...filtres, ...tri, ...pagination } }
  );
  return response.data;
};
